core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -52,50 +65,100 @@
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
-                } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
-                    doCommit(context, transactionManager, transaction);
+                if (readOnly && !isRolledBack) {
+                    try {
+                        doRollback(context, transactionManager, status);
+                    }
+                    finally {
+                        isRolledBack = true;
+                    }
+                } else if (!isRolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
+                    try {
+                        doCommit(context, transactionManager, status);
+                    }
+                    finally {
+                        isCommitted = true;
+                    }
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!isRolledBack) {
+                        try {
+                            doRollback(context, transactionManager, status);
+                        }
+                        finally {
+                            isRolledBack = true;
+                        }
+                    }
                 } else {
                     Log.debug(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
+                    if (!isRolledBack) {
+                        try {
+                            doRollback(context, transactionManager, status);
+                        }
+                        finally {
+                            isRolledBack = true;
+                        }
+                    }
                 }
             } catch (Throwable t) {
                 Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", t);
-                doRollback(context, transactionManager, transaction);
+                if (!isRolledBack) {
+                    try {
+                        doRollback(context, transactionManager, status);
+                    }
+                    finally {
+                        isRolledBack = true;
+                    }
+                }
+            }
+            Log.debug(
+                Log.JEEVES,
+                "Run in transaction completed:" +
+                    (isRolledBack?" rolled back":"") +
+                    (isCommitted?" committed":"")
+            );
+            if( isRolledBack == false && isCommitted == false ){
+                Log.warning(  Log.JEEVES, "Run in transaction did not complete cleanly, transaction not committed or rolledback");
```

Comment:
Can you please consider merging / squashing with https://github.com/cmangeat/core-geonetwork/tree/transaction_manager.

It may help understand that if no commit or rollback it can be because 

(!isNewTransaction) && (commitBehavior != CommitBehavior.ALWAYS_COMMIT)

Reply:
I quite enjoyed your changes, with more smaller methods (much closer to my preferred coding style). I was trying to mirror as much as possible the conventions of these classes in the hopes of making the changes easier to review.

One small bit of feedback, one change seems problematic:
```
                    try {
                        doRollback(context, transactionManager, status);
                    }
                    finally {
                        isRolledBack = true;
                    }
```

replaced with:
```
                    isRolledBack = true;
                    doRollback(context, transactionManager, status);
```

I like to have fields such that make assertions about state be correct at all times in the debugger. I believe your later refactoring got around this and the result should be fine.

I am pushing up all your commits to this branch as is, I think it would be best if this PR is squashed and merged when it is accepted onto master and backported.